config/interfaces.go

```diff
@@ -0,0 +1,15 @@
+package config
+
+// InterfacesConfig configuration for blocky's interfaces (REST, DoH, ...)
+type InterfacesConfig struct {
+	Rest RESTConfig `yaml:"rest"`
+	DoH  DoHConfig  `yaml:"dns-over-http"`
+}
+
+type RESTConfig struct {
+	Enable bool `yaml:"enable" default:"true"`
+}
+
+type DoHConfig struct {
+	Enable bool `yaml:"enable" default:"true"`
+}
```

Comment:
It would be nice to be able to make each HTTP service use a different port: that way you can, for instance, expose DoH on all interfaces but the API only on localhost, and use a firewall as a line of defense without a reverse proxy.
I think that would need more code changes though since there would need to be one `http.Server` for each port, so that can definitely be a follow-up.

Reply:
Agreed 👍

That would be the cleanest and most secure approach.

Considering changing to an port<->service mapping the config options should move to the ports section or maybe create a new service section? 🤔

Reply:
👍 
Definition of port per services would make this PR obsolete. User can decide, which service should be provided by defining the port, otherwise the service is disabled. 
I think it should also be possible to use the same port for different services. Or should we force to use different ports? 🤔 

`http` configuration will not be used anymore. We have also `https` config property, I think it is too much to provide http and https ports for each service, since only for DoH is https mandatory. I would suggest to use https for DoH automatically and http for other services, but it could be complicated if user wants to use the same port for DoH and REST API.

Something like this: 

oldConfig:
```yaml
ports:
  dns: 53
  tls: 853
  http: 4000
  https: 443
```

new config with different ports
```yaml
ports:
  dns: 53 # plain
  tls: 853 # encrypted
  dns-over-https: 443 # encrypted
  prometheus: 4001 # plain http
  control-api: 4000 # plain http
```

new config with same port (like old behavior)
```yaml
ports:
  dns: 53 # plain
  tls: 853 # encrypted
  dns-over-https: 443 # encrypted
  prometheus: 4001 # plain http
  control-api: 4000 # plain http
```

index.html (links to services and version number) will be removed, instead we can provide a new REST-endpoint in control-API with version number.

Any thoughts?

Reply:
Not sure if we actually want this, but maybe allowing prometheus and the API to use HTTPS would be useful?  
Config could look something like:
```yaml
ports:
  dns:
    plain: 53
    tls: 853
    https: 443
  metrics:
    http: 4001
    https: 443 # would use the same router and server as DoH
  control-api:
    http: 4000
```

Having a section unspecified would mean disabled.  
The old options would need to be migrated to set multiple values which is doable with `Apply`. And we'd probably need to remove the old migration because it would not work consistently since it would depend on the older migration to run before the newer one, but iterating on a `map` is unordered so that would not be guaranteed.

> but it could be complicated if user wants to use the same port for DoH and REST API

Yeah the code would need to re-use the `*chi.Mux` per port+proto combo.  
So maybe build a `map[struct{port, protocol}]*chi.Mux` and then loop over that to create each `net.Listener` and server.

> index.html (links to services and version number) will be removed, instead we can provide a new REST-endpoint in control-API with version number.

The REST-endpoint sounds good, and I have no opinion on `index.html`.

Reply:
I like the approach with multiple ports per service config wise(clear structure & extensible).
But is there really a scenario where https connection to the metrics or control API is needed?
Since they aren't secured with credentials or API keys they shouldn't be exposed outside of a secured network anyway...🤔

Reply:
Yeah might not be really useful to add `https` options. But even without that, having the protocol in the option name would be future proof and make it more explicit IMO.

Reply:
Adding quick later on might be interesting. 🤔

Reply:
There are systems for auto creating certificates in multi container environments which could handle it for you (kubernetes has something build in).

I think unsecured + option to add more is fine.

Reply:
> I like the approach with multiple ports per service config wise(clear structure & extensible). But is there really a scenario where https connection to the metrics or control API is needed? Since they aren't secured with credentials or API keys they shouldn't be exposed outside of a secured network anyway...🤔

Maybe we should rename the config from "ports" to "interfaces" and use it for further interface configs (api-key, ...). We can also move the metrics config (prometheus.enable and prometheus.path) there. Something like:

```yaml
interfaces:
  dns:
    ports:
       plain: 53
       tls: 853
       https: 443
  rate-limit: 60 # just example for further config
  metrics:
    ports:
       http: 4001
       https: 443 
    path: /mymetrics # moved from prometheus.path
  control-api:
    ports:
       http: 3000
    api-key: ASBC123 # just example for further config
```

Reply:
I like this structure the most so far.

Reply:
I like it! Just think `interfaces` could be confusing with "network interfaces" meaning something else. What about `listeners` or `servers`?

Reply:
Agreed with the confusing part.

How about `services`?

Technically they are services blocky provides...🤔

Reply:
I created #1206 to track new implementation. I'll close this PR